### ISSUE TITLE
Decoding Hashed Ids in the Request before validation.

### ISSUE THREAD

**Issue by justinatack** [rebing/graphql-laravel#484]
### Versions:
- graphql-laravel Version: 2.1.1
- Laravel Version: 5.6
- PHP Version: 7.3.9

### Question: 

I'm Hashing Id's to obsurce ID's on the frontend however I'm struggling to workout how to hook into the request cycle before the "rules" function and decode the hashed ID. If I comment out the "rules" validation function then updating the resource work fine. The validation error is as follows and shows that the hashed string is reaching the validation rules but it needs to be decode back to an integer.

```
"validation": {
  "id": [
    "The id must be an integer."
  ]
}
```
This is the mutation below.
```
class UpdateLanguageMutation extends Mutation
{
    use HashIdTrait;

    protected function rules(array $args = []): array
    {
        return [
            'id' => ['required', 'integer', 'exists:languages,id'],
            'locale' => ['string', 'size:2', 'unique:languages,locale'],
            'active' => ['boolean'],
            'name' => ['string'],
        ];
    }

    public function resolve($root, $args, $context, ResolveInfo $resolveInfo, Closure $getSelectFields)
    {
        $fields = $getSelectFields();
        $select = $fields->getSelect();
        $with = $fields->getRelations();

        return LanguageRepository::update($args, $this->decode($args['id']));
    }
}
```

I have successfully worked out how to hash the ID in the response. I have created a Type with a resolveIdField function.

```
// LanguageType extends GraphQLType

use HashIdTrait;

protected function resolveIdField($fields, $args)
{
    return $this->encode($fields->id);
}
```

So in summary can somebody help me workout how to modify the incoming hashed string and decode it back to an integer for validation.

**Comment by mfn** [comment]
Usually for IDs, `Type::id()` is used which defaults to `\GraphQL\Type\Definition\IDType`.

Now this `IDType` is a `ScalarType`; and you can make _custom_ scalar types (`artisan make:graphql:scalar` gives you a nice skeleton).

And with scalars, you control exactly how their external representation (GraphQL consumer) and the internal should look like. A scalar type implements the `\GraphQL\Type\Definition\LeafType` (the interface name feels confusing in this context…) which requires:
- `serialize` => how you want the type look like to the consumer of GraphQL
- `parseValue` and `parseLiteral` => how you translate them back from the serialized format from a GraphQL query variable or the GraphQL query respectively

This also means that you this layer is processed **before** it comes to the validation; at least I might assume so.

To give a fictional example, you should be able to call e.g. `encrypt()` in `serialize` and to the reverse with `decrypt()` in the other methods.

To give another example: Scalars can be used to transform forth and back date time stamps. Obviously in JSON there's no "native" type for this, only string. But if you know what types you have, you surely want to work server-side with something more idiomatic than a string, say a `Carbon\Carbon` instance. And scalars are perfectly fit to serialize them to a suitable JSON string representation but also convert them back to Carbion when received.

The same principle would apply to any process where you can use two-way function for your IDs I would assume.

HTH

_PS: technical nitpick: if you call it "hashing" then it's supposed to be "one-way", there's no "decoding of hashes" unless you talk about brute-forcing, rainbow tables, etc._

**Comment by mfn** [comment]
Closing due to lack of further feedback
